src/GeniusYield/TxBuilder/Class.hs

```diff
@@ -77,213 +95,286 @@ module GeniusYield.TxBuilder.Class
+import           GeniusYield.TxBuilder.User
 import           GeniusYield.Types
+import           GeniusYield.Types.Key.Class  (ToShelleyWitnessSigningKey)
 import           GHC.Stack                    (withFrozenCallStack)
 import qualified PlutusLedgerApi.V1           as Plutus (Address, DatumHash,
                                                          FromData (..),
                                                          PubKeyHash, TokenName,
                                                          TxOutRef, Value)
 import qualified PlutusLedgerApi.V1.Value     as Plutus (AssetClass)
 
+-- | Class of monads for building transactions. This can be default derived if the requirements are met.
+-- Specifically, set 'TxBuilderStrategy' to 'GYCoinSelectionStrategy' if you wish to use the default in-house
+-- transaction building implementation.
+class (GYTxSpecialQueryMonad m, GYTxUserQueryMonad m, Default (TxBuilderStrategy m)) => GYTxBuilderMonad m where
+    type TxBuilderStrategy m :: Type
+    type TxBuilderStrategy m = GYCoinSelectionStrategy
+    buildTxBodyWithStrategy :: forall v. TxBuilderStrategy m -> GYTxSkeleton v -> m GYTxBody
+    default buildTxBodyWithStrategy :: forall v. (MonadRandom m, TxBuilderStrategy m ~ GYCoinSelectionStrategy)
+                                    => TxBuilderStrategy m
+                                    -> GYTxSkeleton v
+                                    -> m GYTxBody
+    buildTxBodyWithStrategy = buildTxBodyWithStrategy'
+
+buildTxBody :: GYTxBuilderMonad m => GYTxSkeleton v -> m GYTxBody
+buildTxBody = buildTxBodyWithStrategy def
+
+-- | Class of monads for interacting with the blockchain using transactions.
+class GYTxBuilderMonad m => GYTxMonad m where
+    -- | Sign a transaction body with the user payment key to produce a transaction with witnesses.
+    --
+    -- /Note:/ The key is not meant to be exposed to the monad, so it is only held
+    -- within the closure that signs a given transaction.
+    -- It is recommended to use 'signGYTxBody' and similar to implement this method.
+    signTxBody :: GYTxBody -> m GYTx
+
+    -- | Sign a transaction body with the user payment key AND user stake key to produce
+    -- a transaction with witnesses.
+    -- If the user wallet does not have a stake key, this function should be equivalent to
+    -- 'signTxBody'.
+    --
+    -- See note on 'signTxBody'
+    signTxBodyWithStake :: GYTxBody -> m GYTx
+
+    -- | Submit a fully built transaction to the chain.
+    --   Use 'buildTxBody' to build a transaction body, and 'signGYTxBody' to
+    --   sign it before submitting.
+    --
+    -- /Note:/ Changes made to the chain by the submitted transaction may not be reflected immediately,
+    -- see 'awaitTxConfirmed'.
+    --
+    -- /Law:/ 'someUTxO' calls made after a call to 'submitTx' may return previously returned UTxOs
+    -- if they were not affected by the submitted transaction.
+    submitTx :: GYTx -> m GYTxId
```

Comment:
I wonder whether we should put `submitTx` and `awaitTxConfirmed'` to some other monad as the current structures assumes the requirement of a signing key but what if built transaction doesn't require signing by such a key.

Reply:
Good point. Yes, the monad you're looking for is `GYTxBuilderMonad`. That's essentially the replacement for _current_ `GYTxMonad`. See txmonad hierarchy and migration guide above and also these two comments:

https://github.com/geniusyield/atlas/pull/322#discussion_r1674362687

https://github.com/geniusyield/atlas/pull/322#discussion_r1674370425

I think for app-like usecases (most of GY codebase from what I remember) will just switch to  `GYTxBuilderMonad m => ...` and `runGYTxBuilderMonadIO` and that's pretty much it. That won't require the keys.